File: lib/java/src/main/java/io/github/brenbar/uapi/internal/ServerBinaryEncode.java
package io.github.brenbar.uapi.internal;

import java.util.List;
import java.util.Map;
import java.util.Objects;

import static io.github.brenbar.uapi.internal.EncodeBody.encodeBody;
import static io.github.brenbar.uapi.internal.PackBody.packBody;

public class ServerBinaryEncode {
    static List<Object> serverBinaryEncode(List<Object> message, BinaryEncoding binaryEncoder) {
        final var headers = (Map<String, Object>) message.get(0);
        final var messageBody = (Map<String, Object>) message.get(1);
        final var clientKnownBinaryChecksums = (List<Integer>) headers.remove("_clientKnownBinaryChecksums");

        if (clientKnownBinaryChecksums == null || !clientKnownBinaryChecksums.contains(binaryEncoder.checksum)) {
            headers.put("enc_", binaryEncoder.encodeMap);
        }

        headers.put("bin_", List.of(binaryEncoder.checksum));
        final var encodedMessageBody = encodeBody(messageBody, binaryEncoder);

        final Map<Object, Object> finalEncodedMessageBody;
        if (Objects.equals(true, headers.get("_pac"))) {
            finalEncodedMessageBody = packBody(encodedMessageBody);
        } else {
            finalEncodedMessageBody = encodedMessageBody;
        }

        return List.of(headers, finalEncodedMessageBody);
    }
}

File: lib/ts/uapi/internal/binary/ServerBinaryEncode.ts
import { BinaryEncoding } from 'uapi/internal/binary/BinaryEncoding';
import { encodeBody } from 'uapi/internal/binary/EncodeBody';
import { packBody } from 'uapi/internal/binary/PackBody';

export function serverBinaryEncode(message: any[], binaryEncoder: BinaryEncoding): any[] {
    const headers: { [key: string]: any } = message[0];
    const messageBody: { [key: string]: any } = message[1];
    const clientKnownBinaryChecksums: number[] | undefined = headers['_clientKnownBinaryChecksums'];

    if (clientKnownBinaryChecksums === undefined || !clientKnownBinaryChecksums.includes(binaryEncoder.checksum)) {
        headers['enc_'] = binaryEncoder.encodeMap;
    }

    headers['bin_'] = [binaryEncoder.checksum];
    const encodedMessageBody = encodeBody(messageBody, binaryEncoder);

    let finalEncodedMessageBody: { [key: string]: any };
    if (headers['_pac'] === true) {
        finalEncodedMessageBody = packBody(encodedMessageBody);
    } else {
        finalEncodedMessageBody = encodedMessageBody;
    }

    return [headers, finalEncodedMessageBody];
}
